src/UserAgreement/Quiz.jsx

```diff
@@ -0,0 +1,193 @@
+import './Quiz.less';
+
+const getQuestionSectionId = idx => `Q${idx}`;
+
+const makeDefaultState = (questions) => {
+  const notDone = [];
+  questions.forEach(
+    (item, i) => {
+      notDone.push(i);
+    },
+  );
+  return {
+    notDone,
+    rights: [],
+    wrongs: [],
+    currentSection: getQuestionSectionId(1),
+  };
+};
+
+/**
+ * Little quiz component - roperites: questionList, title, onSubmit
+ */
+class Quiz extends Component {
+  static propTypes = {
+    questionList: PropTypes.arrayOf(
+      PropTypes.shape(
+        {
+          name: PropTypes.string,
+          question: PropTypes.string,
+          options: PropTypes.arrayOf(PropTypes.string),
+          answer: PropTypes.number,
+          hint: PropTypes.string,
+        },
+      ),
+    ),
+    title: PropTypes.string.isRequired,
+    description: PropTypes.string.isRequired,
+    onSubmit: PropTypes.func.isRequired,
+  };
+
+  static defaultProps = {
+    questionList: [],
+  };
+
+  constructor(props) {
+    super(props);
+    this.qs = {};
+    this.qsDom = {};
+    this.state = makeDefaultState(this.props.questionList);
+    this.scrolledByClick = false;
+  }
+
+  componentDidMount() {
+    this.qsList.addEventListener('scroll',
+      () => this.handleScroll(this.qsList));
+    const currentQ = this.qsDom[getQuestionSectionId(Object.values(this.qsDom).length)];
+    this.lastOffsetTop = currentQ.offsetTop;
+  }
+
+  componentWillUnmount() {
+    this.qsList.removeEventListener('scroll',
+      () => this.handleScroll(this.qsList));
+  }
+
+  setCurrentSection(sectionIdx) {
+    this.setState({ currentSection: getQuestionSectionId(sectionIdx + 1) });
+  }
+
+  resetState = () => {
+    this.setState(makeDefaultState(this.props.questionList));
+  };
+
+  answerQuestion(questionId, isCorrect) {
+    const notDone = this.state.notDone.filter(item => item !== questionId);
+    let { rights } = this.state;
+    let { wrongs } = this.state;
+    if (isCorrect) {
+      if (!rights.includes(questionId)) {
+        rights.push(questionId);
+        rights.sort();
+      }
+      wrongs = wrongs.filter(item => item !== questionId);
+    } else {
+      if (!wrongs.includes(questionId)) {
+        wrongs.push(questionId);
+        wrongs.sort();
+      }
+      rights = rights.filter(item => item !== questionId);
+    }
+    this.setState({ notDone, rights, wrongs });
+  }
+
+  handleScroll(ctrl) {
+    if (this.scrolledByClick) {
+      this.scrolledByClick = false;
+      return;
+    }
+    const curPos = ctrl.scrollTop + ctrl.offsetTop;
+    // Get id of current scroll item
+    let { lastOffsetTop } = this;
+    let firstActiveIdx = this.state.currentSection;
+    Object.values(this.qs).forEach((item) => {
+      const dom = this.qsDom[getQuestionSectionId(item.props.idx + 1)];
+      if (dom.offsetTop >= curPos) {
+        if (dom.offsetTop <= lastOffsetTop &&
+          curPos + ctrl.offsetHeight >= dom.offsetTop + (dom.offsetHeight * 0.9)) {
```

Comment:
This `handleScroll` function is difficult for me to understand, I think it'd be great to add more comments / explanations so that others could easily understand its logic, especially this `if` statement (e.g. what does `0.9` mean). Also could we have better name for `ctrl`? for example `scrollController`...